File: python/day09.py
import os
import sys
from typing import List, Optional, Union


class FilePage:
    file_id: int
    file_size: int
    space_after: int

    def __init__(self, file_id: int, file_size: int, space_after: int = 0):
        self.file_id = file_id
        self.file_size = file_size
        self.space_after = space_after

    def __repr__(self) -> str:
        return str(self.file_id) * self.file_size + '.' * self.space_after


def read_data(file_name: str) -> str:
    input_data: str = ""
    with open(file_name, "r") as file:
        return file.read().strip(" \r\n\t")
    return input_data


def uncompress_memory(data: str) -> List[str]:
    memory: List[str] = []
    file_id: int = 0
    isFile: bool = True
    for char in data:
        if (isFile):
            memory += ([f"{file_id}"] * int(char))
            isFile = False
            file_id += 1
        else:
            memory += (["."] * int(char))
            isFile = True
    return memory


def calculate_checksum(memory: Union[List[str], List[FilePage]]) -> int:
    checksum: int = 0
    if memory is not None and all(isinstance(file_page, FilePage) for file_page in memory):
        idx: int = 0
        for file_page in memory:
            start_idx = idx
            for i in range(start_idx, start_idx + file_page.file_size):
                checksum += idx * file_page.file_id
                idx += 1
            idx += file_page.space_after
    elif memory is not None and all(isinstance(file_page, str) for file_page in memory):
        first_blank = memory.index(".")
        for idx, value in enumerate(memory[:first_blank]):
            checksum += idx * int(value)
    else:
        raise Exception("InvalidType")
    return checksum


def part1(memory: List[str]) -> int:
    start: int = 0
    end: int = len(memory)-1
    while start < end:
        while memory[start] != ".":
            start += 1
        while not memory[end].isdigit():
            end -= 1
        if start >= end:
            break
        temp = memory[start]
        memory[start] = memory[end]
        memory[end] = temp
    return calculate_checksum(memory)


def construct_file_pages(data: str) -> List[FilePage]:
    file_pages: List[FilePage] = []
    file_id: int = 0
    for idx, char in enumerate(data):
        file: FilePage = None
        if idx % 2 == 0:
            file = FilePage(file_id, int(char))
            file_pages.append(file)
            file_id += 1
        else:
            file_pages[-1].space_after = int(char)
    return file_pages


def part2(file_pages: List[FilePage]) -> int:
    end = len(file_pages) - 1
    while end > 0:
        start = 0
        while file_pages[start].space_after < file_pages[end].file_size and start < end:
            start += 1
        if start >= end:
            end -= 1
            continue
        end_page = file_pages[end]
        file_pages[end-1].space_after += file_pages[end].file_size + \
            file_pages[end].space_after
        end_page.space_after = file_pages[start].space_after - \
            end_page.file_size
        file_pages.remove(file_pages[end])
        file_pages[start].space_after = 0
        file_pages.insert(start+1, end_page)

    return calculate_checksum(file_pages)


if __name__ == "__main__":
    input_file: str = "/home/mohitjangra/learning/advent_of_code_2024/inputs/day09.input"
    test_file: str = "/home/mohitjangra/learning/advent_of_code_2024/tests/day09.test"

    mode = os.sys.argv[1] if len(os.sys.argv) > 1 else "test"
    if mode.strip().lower() == "input":
        compressed_memory = read_data(input_file)

        uncompressed = uncompress_memory(compressed_memory)
        pt1_result = part1(uncompressed)
        assert pt1_result == 6330095022244

        file_pages: List[FilePage] = construct_file_pages(compressed_memory)
        pt2_result = part2(file_pages)
        assert pt2_result == 6359491814941
    elif mode.strip().lower() == "test":
        compressed_memory = read_data(test_file)

        uncompressed = uncompress_memory(compressed_memory)
        assert "".join(
            uncompressed) == "00...111...2...333.44.5555.6666.777.888899"
        pt1_result = part1(uncompressed)
        assert pt1_result == 1928

        file_pages: List[FilePage] = construct_file_pages(compressed_memory)
        assert "".join(
            [str(page) for page in file_pages]) == "00...111...2...333.44.5555.6666.777.888899"
        pt2_result = part2(file_pages)
        assert pt2_result == 2858
    else:
        print(f"Usage: {os.sys.argv[0]} [test|input]", file=sys.stderr)

File: zig/day09.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

const File = struct {
    file_id: usize,
    file_size: usize,
    space_after: usize,
    start_idx: usize = 0,
};

fn read_data(allocator: Allocator, filename: []const u8) ![]const u8 {
    const file = try std.fs.openFileAbsolute(filename, .{ .mode = .read_only });
    defer file.close();

    const file_stats = try file.stat();
    const buffer: []u8 = try allocator.alloc(u8, file_stats.size);
    const bytes_read = try file.read(buffer);
    std.debug.assert(file_stats.size == bytes_read);
    return buffer;
}

fn parse_memory(allocator: Allocator, compressed: []const u8) !std.DoublyLinkedList(File) {
    var list = std.DoublyLinkedList(File){};
    var is_file: bool = true;
    var file_id: u32 = 0;
    var start_idx: usize = 0;
    for (std.mem.trim(u8, compressed, " \r\n\t"), 0..) |char, idx| {
        if (is_file) {
            const size: u8 = try std.fmt.parseInt(u8, &[_]u8{char}, 10);
            const node = try allocator.create(std.DoublyLinkedList(File).Node);
            node.data = File{
                .file_id = file_id,
                .file_size = size,
                .start_idx = start_idx,
                .space_after = 0,
            };
            list.append(node);
            file_id += 1;
            start_idx += size;
        } else {
            const space: u8 = try std.fmt.parseInt(u8, &[_]u8{compressed[idx]}, 10);
            if (list.last) |item| {
                item.data.space_after = space;
            }
            start_idx += space;
        }
        is_file = !is_file;
    }
    return list;
}

fn uncompress(allocator: Allocator, data: []const u8) ![]i32 {
    var memory_size: usize = 0;
    for (std.mem.trim(u8, data, " \r\n\t")) |char| {
        memory_size += try std.fmt.parseInt(u8, &[_]u8{char}, 10);
    }
    const buffer: []i32 = try allocator.alloc(i32, memory_size);
    var is_file: bool = true;
    var file_id: u32 = 0;
    var idx: usize = 0;
    for (std.mem.trim(u8, data, " \r\n\t")) |char| {
        var count: u8 = try std.fmt.parseInt(u8, &[_]u8{char}, 10);
        if (is_file) {
            while (count > 0) {
                buffer[idx] = @intCast(file_id);
                count -= 1;
                idx += 1;
            }
            file_id += 1;
        } else {
            while (count > 0) {
                buffer[idx] = -1;
                count -= 1;
                idx += 1;
            }
        }
        is_file = !is_file;
    }
    return buffer;
}

fn print_memory(T: type, memory: T) !void {
    switch (T) {
        std.DoublyLinkedList(File) => {
            var it: ?*std.DoublyLinkedList(File).Node = memory.first;
            while (it) |item| : (it = item.next) {
                for (0..item.data.file_size) |_| {
                    std.debug.print("{d}", .{item.data.file_id});
                }
                for (0..item.data.space_after) |_| {
                    std.debug.print(".", .{});
                }
            }
            std.debug.print("\n", .{});
        },
        []i32 => {
            for (memory) |item| {
                if (item < 0) {
                    std.debug.print(".", .{});
                } else {
                    std.debug.print("{d}", .{item});
                }
            }
            std.debug.print("\n", .{});
        },
        else => {
            return error.InvalidType;
        },
    }
}

fn calculate_checksum(comptime T: type, memory: T) !usize {
    var checksum: usize = 0;
    switch (T) {
        std.DoublyLinkedList(File) => {
            var it: ?*std.DoublyLinkedList(File).Node = memory.first;
            while (it) |item| : (it = item.next) {
                for (item.data.start_idx..item.data.start_idx + item.data.file_size) |idx| {
                    checksum += idx * item.data.file_id;
                }
            }
        },
        []i32 => {
            for (memory, 0..) |item, idx| {
                if (item == -1) continue;
                checksum += @as(usize, @intCast(item)) * @as(usize, @intCast(idx));
            }
        },
        else => {
            return error.InvalidType;
        },
    }
    return checksum;
}

fn part1(memory: []i32) !usize {
    var start: usize = 0;
    var end: usize = memory.len - 1;

    while (end > start) {
        while (memory[end] == -1) {
            end -= 1;
        }
        while (memory[start] != -1) {
            start += 1;
        }
        if (start < end) {
            std.mem.swap(i32, &memory[start], &memory[end]);
        }
    }
    return try calculate_checksum([]i32, memory);
}

fn part2(memory: *std.DoublyLinkedList(File)) !usize {
    var end = memory.last;
    var temp_end_node: ?*std.DoublyLinkedList(File).Node = undefined;
    while (end) |node| : (end = temp_end_node) {
        temp_end_node = node.prev;
        var start = memory.first;
        while (start) |space| : (start = space.next) {
            if (space == node) break;
            if (space.data.space_after >= node.data.file_size) {
                const temp = node;
                memory.remove(node);
                temp.prev.?.data.space_after += temp.data.file_size + temp.data.space_after;
                temp.data.space_after = space.data.space_after - temp.data.file_size;
                temp.data.start_idx = space.data.start_idx + space.data.file_size;
                memory.insertAfter(space, temp);
                space.data.space_after = 0;
                break;
            }
        }
    }
    return try calculate_checksum(std.DoublyLinkedList(File), memory.*);
}

pub fn main() !void {
    const allocator = std.heap.page_allocator;
    const filename = "/home/mohitjangra/learning/advent_of_code_2024/inputs/day09.input";

    const buffer: []const u8 = try read_data(allocator, filename);
    defer allocator.free(buffer);

    const memory = try uncompress(allocator, buffer);
    defer allocator.free(memory);
    const memory_copy = try uncompress(allocator, buffer);
    defer allocator.free(memory_copy);

    const pt1_result = try part1(memory);
    std.debug.assert(pt1_result == 6330095022244);

    var list = try parse_memory(allocator, buffer);
    defer {
        var it = list.first;
        var next: ?*std.DoublyLinkedList(File).Node = undefined;
        while (it) |item| : (it = next) {
            next = item.next;
            allocator.destroy(item);
        }
    }
    const pt2_result = try part2(&list);
    std.debug.assert(pt2_result == 6359491814941);
}

test {
    const allocator = std.testing.allocator;
    const filename = "/home/mohitjangra/learning/advent_of_code_2024/tests/day09.test";
    const buffer: []const u8 = try read_data(allocator, filename);
    defer allocator.free(buffer);

    const memory = try uncompress(allocator, buffer);
    defer allocator.free(memory);

    const memory_copy = try uncompress(allocator, buffer);
    defer allocator.free(memory_copy);
    const pt1_result = try part1(memory);
    try std.testing.expect(pt1_result == 1928);

    var list = try parse_memory(allocator, buffer);
    defer {
        var it = list.first;
        var next: ?*std.DoublyLinkedList(File).Node = undefined;
        while (it) |item| : (it = next) {
            next = item.next;
            allocator.destroy(item);
        }
    }
    const pt2_result = try part2(&list);
    try std.testing.expect(pt2_result == 2858);
}

test "file linked list" {
    const allocator = std.testing.allocator;
    const filename = "/home/mohitjangra/learning/advent_of_code_2024/tests/day09.test";
    const buffer: []const u8 = try read_data(allocator, filename);
    defer allocator.free(buffer);

    var list = try parse_memory(allocator, buffer);
    defer {
        var it = list.first;
        var next: ?*std.DoublyLinkedList(File).Node = undefined;
        while (it) |item| : (it = next) {
            next = item.next;
            allocator.destroy(item);
        }
    }
    const pt2_result = try part2(&list);
    try std.testing.expect(pt2_result == 2858);
}
